src/documentation/programming-guide.md

```diff
@@ -900,8 +901,190 @@ odds = results.odd
+
+#### Writing to a `Sink`:
+
+```java
+// This example uses XmlSink.
+pipeline.apply(Write.to(
+          XmlSink.ofRecordClass(Type.class)
+              .withRootElementName(root_element)
+              .toFilenamePrefix(output_filename)));
+```
+
+```python
+output | beam.io.Write(beam.io.TextFileSink('my_file_name'))
+```
+
+#### Using a write transform:
+
+```java
+// This example uses JmsIO.
+pipeline.apply(...) // returns PCollection<String>
+        .apply(JmsIO.write()
+            .withConnectionFactory(myConnectionFactory)
+            .withQueue("my-queue")
+```
+
+```python
+output | beam.io.textio.WriteToText('my_file_name')
+```
+
+### File-based input and output data
+
+#### Reading From Multiple Locations:
+
+Many read transforms support reading from multiple input files matching a glob operator you provide. The following TextIO example uses a glob operator (\*) to read all matching input files that have prefix "input-" and the suffix ".csv" in the given location:
+
+```java
+p.apply(“ReadFromText”,
+    TextIO.Read.from("protocol://my_bucket/path/to/input-*.csv");
+```
+
+```python
+lines = p | beam.io.Read(
+    'ReadFromText',
+    beam.io.TextFileSource('protocol://my_bucket/path/to/input-*.csv'))
+```
+
+To read data from disparate sources into a single `PCollection`, read each one independently and then use the `Flatten` transform to create a single `PCollection`.
+
+#### Writing To Multiple Output Files:
+
+For file-based output data, write transforms write to multiple output files by default. When you pass an output file name to a write transform, the file name is used as the prefix for all output files that the write transform produces. You can append a suffix to each output file by specifying a suffix.
+
+The following write transform example writes multiple output files to a location. Each file has the prefix "numbers", a numeric tag, and the suffix ".csv".
+
+```java
+records.apply(TextIO.Write.named("WriteToText")
+                          .to("protocol://my_bucket/path/to/numbers")
+                          .withSuffix(".csv"));
+```
+
+```python
+filtered_words | beam.io.Write(
+    'WriteToText',
+    beam.io.TextFileSink('protocol://my_bucket/path/to/numbers',
+                                        file_name_suffix='.csv'))
+```
+
+### Beam provided I/O APIs
+
+See the language specific source code directories for the Beam supported I/O APIs. Specific documentation for each of these I/O sources will be added in the future. ([BEAM-1054](https://issues.apache.org/jira/browse/BEAM-1054))
+
+<table class="table table-bordered">
+<tr>
+  <th>Language</th>
+  <th>File-based</th>
+  <th>Messaging</th>
+  <th>Database</th>
+</tr>
+<tr>
+  <td>Java</td>
+  <td>
+    <p><a href="https://github.com/apache/incubator-beam/tree/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io">Avro</a></p>
```

Comment:
lots of repetition of the phrase `https://github.com/apache/incubator-beam/tree/master/`. It would be good to move it into a variable.